[WIP]integrated with Virtualcodebase of scancode(First Approach)

pratik0316 opened the pull request:
In this PR I integrated the virtualcodebases of Scancode to the Deltacode.

pratik0316 committed:
integrated with Virtualcodebase of scancode

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
fixed for test cases tests/test_deltacode.py::TestDeltacode::test_score_no_lic_change and test_Delta_one_None

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
passed for test_score_no_copyright _no_copyright_changes

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
changed the passing parameter to_dict()

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -48,14 +49,27 @@ class DeltaCode(object):
+            self.codebase1 = VirtualCodebase(new_path)
+            self.codebase2 = VirtualCodebase(old_path)
+        except :
+            pass
```
We want to handle the exception properly

pratik0316 replied:
@majurg okay.
The exceptions were raised when we are getting some invalid sacn paths , and when we counter some some attributes like "fiingerprint" which VirtualCodebase is not supporting

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -48,14 +49,27 @@ class DeltaCode(object):
+            self.codebase2 = VirtualCodebase(old_path)
+        except :
+            pass
+        self.new_files_count = 0 #keeps the count of the new file
+        self.old_files_count = 0 #keeps the count of old files
+        self.new_files = [] # a list of [[new file1:Original path],[new file2:Original Path],...]
+        self.old_files = [] # a list of [[old file1:Original path],[old file2:Original Path],...]
+        self.new_files_fingerprint = dict() # map of { {new_file1:fingerprint},{new_file2:fingerprint},...} it will be needed when we need the fingerprints
+        self.old_files_fingerprint = dict() # map of { {old_file1:fingerprint},{old_file2:fingerprint},...}
```
What is all this stuff and why is it needed?

pratik0316 replied:
`self.new_files_count and self.old_files_count` is to keep the track of the new files ,and old files.
I have added  this to make the computation of the statistics easier,otherwise we would have to enumerate the virtual codebase objects to get the cont every time.
`self.new_files_fingerprint and self.old_files_fingerprint` it keeps a mapping of  Resource objects files  path from `codebase1 and codebase2` with respect to the `fingerprints` which would be used in `similarity` comparisons.

pratik0316 replied:
@majurg  `self.new_files and self.old_files ` it is list of lists comprising of [`Resource objects`,`with their original path`] ,Now we need to keep the track of the original path in the `align_scans` for alignment of the files.

steven-esser replied:
Its fine to enumerate on the Virtualcodebase. Stats should be calculated in the end anyway, and optionally for the user. I hate having carrying around all these unneeded fields.

steven-esser replied:
Also, codebase objects have counts already that we can use. There is no need to track this twice.

pratik0316 replied:
@majurg  , yes we can get rid of this `self.new_files_count` and `self.old_files_count` as they are already present in codebase objects  as it is present in the headers of the `codebase objects`, but for the `old files` and `new files` and the `fingerprint`, I think it is better to have the `enumeration` done one time and cache those files in an array, else we will again need to `enumerate` it whenever required.

pratik0316 replied:
@majurg used the counts from the codebase , removed the additional variables for the `files_count`

pratik0316 committed:
Fixed the issues for some test cases like similarity matching1 and 2 and non similarity matching

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 commented:
@majurg the test cases like (similarity matching1 and similarity matching2 ,non similarity matching1 ) are passing, the output in json file is appearing as expected.
However the code for this kinda test cases can not be modified
```
    def test_Delta_create_object_removed(self):
        new = None
        old = models.File({'path': 'path/removed.txt'})

        delta = deltacode.Delta(0, new, old)
        delta.factors.append('removed')

        assert type(delta.new_file) == type(None)
        assert delta.old_file.path == 'path/removed.txt'
        assert 'removed' in delta.factors
        assert delta.score == 0
``` 
as they are treating the old and new file as File objects and we are using the ScannedResources object for the old and the file.So should I modify test case like this?

pratik0316 committed:
test_deltacode_to_dict_modified

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
fixed for test cases: sending None paths in test_deltacode.py

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
changed the json files which were scanned with version 2 to version 3

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 commented:
@majurg , I replaced some of the json files which were scanned with ScanCode version 2 with approximately similar files scanned with Scancode Version 3 [link](https://github.com/nexB/deltacode/pull/151/files#diff-402375338236ff3c6d7d2966e38c4b57L2) , and modified some of the test cases.
This resulted in passing the pytests for `test_cli.py`.
Kindly review the changes once :)

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -48,14 +50,30 @@ class DeltaCode(object):
+        try:
+            self.codebase1 = VirtualCodebase(new_path)
+            self.codebase2 = VirtualCodebase(old_path)
+        except Exception as exception:
+            click.secho(exception.message ,fg = "red")
```
What exact exceptions can occur here?

pratik0316 replied:
@majurg , the error faced here are following two types:
1.When the paths of the json files for the are not valid.
2.When the VirtualCode base gets an unnecessary fields during the scanning .
However the 2nd error could be resolved if we use scancode versionn3 during the scanning of the json files.

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -48,14 +50,30 @@ class DeltaCode(object):
-
-        if self.new.path != '' and self.old.path != '':
+       
+        if self.codebase1 != None and self.codebase2 != None:
```
Can we just have a check for the inverse of this so we dont write everything under an unneeded indentation?

pratik0316 replied:
@majurg  yeah  , trying for it , I think it would be a better approach

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -67,6 +85,48 @@ def __init__(self, new_path, old_path, options):
+            if is_new:
+                # append in the new_files
+                self.new_files.append([obj,''])
+                try :
+                    self.new_files_fingerprint[obj.path] = obj.fingerprint
+                except AttributeError:
+                    self.new_files_fingerprint[obj.path] = None
+                if obj.is_file:
+                    # increment the new files count
+                    self.new_files_count += 1
```
Why not use `dict.get()` method instead of catching exception?

pratik0316 replied:
@majurg it would be nice  , changing it

pratik0316 replied:
@majurg , the `dict.get()` method can not be used here as `obj` is not a `dictionary` it is a `Virtualcodebase` object.

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -67,6 +85,48 @@ def __init__(self, new_path, old_path, options):
+                    self.old_files_count += 1
+                    
+
+    def enumerate_files_from_codebases(self):
+       """
+       An method which call the utility function get_files for generating the codebase
+       """
+       self.get_files(self.codebase1,is_new = True)
+       self.get_files(self.codebase2,is_new = False)
```
This function's name is unrelated to what it actually does.

pratik0316 replied:
@majurg , I think interchanging the function names `get_files` and `enumerate_files_from_codebases` would better match with the situation

pratik0316 replied:
changed the function name

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -323,7 +388,104 @@ def is_added(self):
+                all_licenses.append(d)
+            return all_licenses
+
+    def new_file_to_dict(self,deltacode):
```
Why is this not a method of the File class?

pratik0316 replied:
@majurg , I will be moving all these methods(`license` and `copyright`) to the `file ` class

steven-esser commented on src/deltacode/__init__.py:
```diff
@@ -323,7 +388,104 @@ def is_added(self):
+            ])
+        
+    
+    def old_file_to_dict(self,deltacode):
```
Why are we repeating this function?

pratik0316 replied:
@majurg , I will be truncating this unnecessary functions.

pratik0316 replied:
combined redundant functions `def old_file_to_dict(self,deltacode)` and `def new_file_to_dict(self,deltacode)` , to a single function.

steven-esser reviewed (changes requested):
There is a lot to be done here before it can be considered for merging.

pratik0316 commented:
Thanks @majurg for  the review , I will be making the suggested changes ASAP :)

pratik0316 committed:
changed redundant functions

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
removed pyvenv.cfg

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
appende the key errors

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
fixed some test cases for tests/test_utils.py

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
added fetch files without creating deltacode object

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 commented on src/deltacode/test_utils.py:
```diff
@@ -168,3 +169,14 @@ def streamline_headers(headers):
+
+
+def fetch_files(location):
+    codebase = VirtualCodebase(location)
```
This function has been extensively added to `test_utils` so as to prevent the creation of `deltacode` object for the test cases when only the `Resource files` are required

pratik0316 committed:
fixed test cases for test_align_tress_zlib_failing

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
used the files count of the VirtualCodebase instead creating a separate one

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
fixed for test_update_from_license_info_copyleft_to_different_copyleft

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
fixed the original paths

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

pratik0316 committed:
few changes in checking conditionals in init.py

Signed-off-by: Pratikrocks <pratikrocks.dey11@gmail.com>

steven-esser commented:
Closing for now due to stagnation.

pratik0316 commented:
I was willing to continue this , due to lack of input I discontinued for a moment , anyway if it's anymore required I will work on it :)

steven-esser commented:
@Pratikrocks Yes this may have fallen through the cracks, sorry about this lack of feedback. 

The deltacode repo has been updated since this PR was made with a number of configuration changes. If you can, please rebase your branch, or add your changes to a new branch off of the current develop `HEAD`. I will make sure to leave prompt feedback if you do!
